src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs

```diff
@@ -457,8 +459,32 @@ impl<'tcx> GotocCtx<'tcx> {
                     let relative_discr = if *niche_start == 0 {
                         niche_val
                     } else {
-                        // This should be a wrapping sub.
-                        niche_val.sub(Expr::int_constant(*niche_start, discr_ty.clone()))
+                        // Compute "niche_val - niche_start" where "-" is
+                        // wrapping subtraction, i.e., the result should be
+                        // interpreted as an unsigned value.
+                        // While this can be done through a regular subtraction
+                        // and then casting the result to an unsigned, doing so
+                        // may result in CBMC flagging the operation with a
+                        // signed to unsigned overflow failure (see
+                        // https://github.com/model-checking/kani/issues/356).
+                        // To avoid those overflow failures, the wrapping
+                        // subtraction operation is computed as:
+                        //   if niche_val >= niche_start {
+                        //       // result is positive, so overflow may not occur
+                        //       niche_val - niche_start
+                        //   } else {
+                        //       // compute the 2's complement to avoid overflow
+                        //       niche_val - niche_start + 2^32
+                        //   }
+                        let s64 = Type::signed_int(64);
+                        let niche_val = niche_val.cast_to(s64.clone());
+                        let twos_complement: i64 =
```

Comment:
I'm adding a general 2's complement in #981

Reply:
Thanks for adding a function for a general 2's complement! This one is for a specific width though, so was simple enough.